### PR TITLE
General inference procedure for ML4H models with applications to segmentations

### PR DESCRIPTION
**Pull request by mklarqvist**
* Load a ML4H/Keras/TF model and make inference on batches of data.
* Compute segmentation masks and store on disk

No inference probabilities are stored in this application

### CODE REVIEW

File: ml4h/applications/segmentation/mri_inference.py [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+from ml4h.normalizer import ZeroMeanStd1
+import numpy as np
+import h5py
+import os
+import cv2
+# import glob # Use glob if a manifest is not available
+import fastparquet as fp
+import sys
+
+# Command line helper for divide-and-conquer approach to inferring
```
Comment (lucidtronix) [inline ml4h/applications/segmentation/mri_inference.py:19]: Wrap this all in a function named `infer_pngs` or something

**Review by lucidtronix — CHANGES_REQUESTED** [review]
Awesome work! Let's integrate this more in the ml4h package by using our arguments namespace and adding it the ml4h package.  Eventually we'll want to add all our inference recipes here.  Besides that little nits about logging and how to get the auxiliary files.

File: ml4h/applications/segmentation/mri_inference.py [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+            # Loop over the last channel in our tensor. This corresponds to the time component
+            for d in range(dat.shape[-1]):
+                try:
+                    # Store the DICOM name
+                    dicom_names.append(str(inner_manifest.loc[manifest['instance_number']==int(d+1)].dicom_file.values[0]))
+                except Exception as e:
+                    print(f'cant find dicom name for: {ukb_id}, {i}, {d}, {instance[20:]}')
+                    continue
+                # Reshape tensor-slice at index d to (224, 224, 1) and store in tensor
+                tensor = pad_or_crop_array_to_shape((224, 224, 1), dat[..., d])
```
Comment (lucidtronix) [inline ml4h/applications/segmentation/mri_inference.py:170]: Can this iterate over `args.tensor_maps_output` and use tm.shape?

File: ml4h/applications/jpp_inference_rv/infer_to_hd5_local.sh [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    do
+    # Attach disk(s) to the spawned VM instance.
+	gcloud compute instances attach-disk $vm_base_name-${i} --disk=$target_disk --mode ro &
+	sleep 1 # Sleep for a second
+    done # End loop i
+    wait # Wait until completeion
+    sleep 25 # Sleep for 25 seconds 
+    for i in $(seq $start $end)
+    do
+	gcloud compute ssh $vm_base_name-${i} --command="cd /home/pdiachil;cd ml;git pull;git checkout pd_sf_blox;git pull;nohup bash /home/pdiachil/ml/scripts/infer_to_hd5.sh $i > /home/pdiachil/out_${i}.out 2> /home/pdiachil/out_${i}.err < /dev/null &" &
```
Comment (lucidtronix) [inline ml4h/applications/jpp_inference_rv/infer_to_hd5_local.sh:35]: can we replace `pdiachil` with `${USER}`

File: ml4h/applications/jpp_inference_rv/infer_to_hd5_local.sh [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    do
+    for i in $(seq $start $end)
+    do
+    # Spawn a VM instance given prepared VM snapshot.
+	gcloud compute instances create $vm_base_name-${i} \
+	       --machine-type=n1-standard-8 \
+	       --boot-disk-size=150 \
+	       --image=$target_vm_snapshot \
+	       --maintenance-policy=TERMINATE \
+	       --accelerator=type=nvidia-tesla-p4,count=1 &
```
Comment (lucidtronix) [inline ml4h/applications/jpp_inference_rv/infer_to_hd5_local.sh:19]: p4 -> t4

**Review by lucidtronix — APPROVED** [review]
2 small changes and then good to merge, no need to bounce back!

**Comment by lucidtronix** [comment]
@mklarqvist back to you

File: ml4h/applications/jpp_inference_rv/infer_to_hd5_local.sh [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+	       --accelerator=type=nvidia-tesla-t4,count=1 &
+	sleep 1  # Sleep for a second
+    done # End loop i
+    wait # Wait until completion
+    done # End loop k
+    wait # Wait until completion
+    for i in $(seq $start $end)
+    do
+    # Attach disk(s) to the spawned VM instance.
+	gcloud compute instances attach-disk ${vm_base_name}-${i} --disk=${target_disk} --mode ro &
```
Comment (paolodi) [inline ml4h/applications/jpp_inference_rv/infer_to_hd5_local.sh:35]: @mklarqvist thanks for fixing these as command line arguments!!!

**Pull request merged**